Issue805 Multizone VAV AHU economizer control sequence

milicag opened the pull request:
This PR adds a multizone VAV AHU economizer control sequence as prescribed by ASHRAE Guideline 36 (G36). The PR comprises 3 atomic and 1 composite control sequence models with the corresponding tests, scripts and validation results. 

Atomic control sequences:

- ```EconDamperPositionLimitsMultiZone```:  Outdoor and return air damper position limits for minimum outdoor airflow control
- ```EconEnableDisableMultiZone```: Economizer enable/disable switch
- ```EconModulationMultiZone```: Outdoor and return air damper position modulation for economizer operation

Composite control sequence:

- ```EconomizerMultiZone```: Multiple zone VAV AHU economizer control sequence

In addition, the PR introduces ```OperationMode```, ```ZoneState```, and ```FreezeProtectionStage``` ```Types```, also according to G36.

milicag commented:
@mwetter It's worth noting that the validation tests are reporting the license issue again, although the reference results did get generated, so I've simply committed those.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/EconModulationMultiZone.mo:
```diff
@@ -0,0 +1,194 @@
+    annotation (Placement(transformation(extent={{-160,40},{-120,80}}),
+        iconTransformation(extent={{-120,-100},{-100,-80}})));
+  CDL.Interfaces.RealInput uRetDamPosMax(min=0, max=1, unit="1")
+    "Maximum return air damper position limit as returned by the EconEnableDisableMultiZone sequence."
```
Don't use a period in the comment string. (Also below are a few comments with periods.)

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/EconModulationMultiZone.mo:
```diff
@@ -0,0 +1,194 @@
+        iconTransformation(extent={{100,10},{120,30}})));
+
+  CDL.Continuous.LimPID damPosController(
+    Td=0.1,
```
I suggest we keep the default values for Td and Nd, unless you have reasons to overwrite them.

milicag replied:
The controller seems to be a bit inconsistent. I am using, per Brent's recommendation, a PI controller. If I set the controller type in ```LimPID``` to PI, both Td and Nd parameters become disabled (as expected). However, as I remove those params from my model, I get an error - only for Td. It turns out Nd has a default value of 10, whereas a default is not set for Td. On the fly I was not able to see why the controller still requires Td to get assigned a value, even though I'm using a PI controller, so ```with_d``` value is 0. For now I'm leaving Td in the code until I get your feedback.

milicag replied:
![pi_cont_default](https://user-images.githubusercontent.com/5322943/27926771-6af54f1c-623e-11e7-90d2-ba1e081be181.PNG)

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/EconModulationMultiZone.mo:
```diff
@@ -0,0 +1,194 @@
+    Nd=1,
+    controllerType=Buildings.Experimental.OpenBuildingControl.CDL.Types.SimpleController.PI,
+    Ti=300,
+    yMax=yConSigMax,
```
yMax, yMin and k should be final.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/EconModulationMultiZone.mo:
```diff
@@ -0,0 +1,194 @@
+  "Minimum control loop signal for the return air damper";
+  parameter Real yConSigMin=0 "Lower limit of controller output";
+  parameter Real yConSigMax=1 "Upper limit of controller output";
+  parameter Real controllerGain=1 "Gain of controller";
```
Please rename `controllerGain` to `k` (or `kP` if it is a proportional gain and you have other gains as well.)

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/EconModulationMultiZone.mo:
```diff
@@ -0,0 +1,194 @@
+  CDL.Continuous.Constant outDamMinLimSig(final k=damPosController.yMin)
+    "Minimal control loop signal for the outdoor air damper."
+    annotation (Placement(transformation(extent={{-20,-20},{0,0}})));
+  CDL.Continuous.Constant outDamMaxLimSig(k=outDamConSigMax)
```
Is there a reason why `k` above is `final`, but this is not final?

milicag replied:
The one above is the controller min signal, which is final. The one below is usually set to 0.5, but that value is not strictly fixed. I could imagine that some economizers could operate better if, for example, we allocate larger fraction of the control signal to outdoor air damper, compared to the return air damper. This is if ```fixed``` should precede a parameter which will never require changes.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/EconModulationMultiZone.mo:
```diff
@@ -0,0 +1,194 @@
+    the supply air control loop signal. This is a final sequence in the 
+    composite multizone VAV AHU economizer control sequence. Damper position 
+    limits, which are the inputs to the sequence, are the outputs of 
+    <code>EconDamperPositionLimitsMultiZone</code> and 
```
Please add a hyperlink, using the full class name, e.g., `<a href=\"modelica://Buildings.Experimental. ...`

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/EconModulationMultiZone.mo:
```diff
@@ -0,0 +1,194 @@
+    Td=0.1,
+    Nd=1,
+    controllerType=Buildings.Experimental.OpenBuildingControl.CDL.Types.SimpleController.PI,
+    Ti=300,
```
The integral time constant `Ti` also needs to be propagated.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/VAVSingleZoneTSupSet.mo:
```diff
@@ -22,7 +21,6 @@ block VAVSingleZoneTSupSet
     annotation (Dialog(group="Speed"));
 
-
   CDL.Interfaces.RealInput uHea(min=0, max=1)
```
Add `unit="1"`

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/VAVSingleZoneTSupSet.mo:
```diff
@@ -500,6 +498,16 @@ economizer dampers.
 based on the same temperature sensors and control loops
 </p>
+<p>
+fixme, mg notes: 
```
Address and remove `fixme`. Once the code is merged, they otherwise won't get address easily and can hold up code releases.

milicag replied:
These notes are some details from the guideline, so I'm simply going to copy them over to my stack of detail related questions. They do not impact the general functionality of the sequence, in the large majority of the cases.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconEnableDisableMultiZone_TOut_hOut.mo:
```diff
@@ -0,0 +1,209 @@
+          textString="Example high limit cutoff conditions:
+                      ASHRAE 90.1-2013:
+                      Device Type: Fixed Enthalpy + Fixed Drybulb, Fixed Drybulb
+                      TOut > 75 degF [24 degC] [297 K]
```
No need to add [297 K]

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconEnableDisableMultiZone_TOut_hOut.mo:
```diff
@@ -0,0 +1,209 @@
+    This example validates
+    <a href=\"modelica://Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.AtomicSequences.EconEnableDisableMultiZone\">
+    Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.AtomicSequences.EconEnableDisableMultiZone</a>
+    for the following control signals: TOut, TOutCut, hOut, hOutCut.
```
Use `<code>` for variable names.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconModulationMultiZone_TSup.mo:
```diff
@@ -0,0 +1,73 @@
+  "Validation model for multiple zone VAV AHU outdoor and return air damper position modulation sequence"
+  extends Modelica.Icons.Example;
+
+  parameter Real TCooSet(unit="K", quantity="TermodynamicTemperature")=291
```
I suggest you use 291.15 so it will be round numbers in Celsius.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconModulationMultiZone_TSup.mo:
```diff
@@ -0,0 +1,73 @@
+  This example validates
+  <a href=\"modelica://Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.AtomicSequences.EconModulationMultiZone\">
+  Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.AtomicSequences.EconModulationMultiZone</a>
+  for supply air temeperature and supply air temperature cooling setpoint (<code>TSup, TCooSet</code>) 
```
Use `... TSup</code>, <code>TCooSet...`

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/package.mo:
```diff
@@ -0,0 +1,20 @@
+within Buildings.Experimental.OpenBuildingControl.ASHRAE.G36;
+package AtomicSequences "Subsequences as defined in guidline G36"
+
```
Remove empty lines

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,179 @@
+within Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.CompositeSequences;
+model EconomizerMultiZone "Multiple zone VAV AHU economizer control sequence"
+
+  parameter Boolean fixEnt = true
```
This would be clearer if called `use_enthalpy` or `have_enthalpyMeasurements`

milicag replied:
Thanks, I was really unhappy with that variable name.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,179 @@
+  parameter Real delEntHis(unit="J/kg", quantity="SpecificEnergy")=1000
+    "Delta between the enthalpy hysteresis high and low limits"
+    annotation(Evaluate=true, Dialog(group="Enthalpy sensor in use", enable = fixEnt));
+  parameter Real delTemHis=1 "Delta between the temperature hysteresis high and low limits";
```
Units are missing

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,179 @@
+    annotation(Evaluate=true, Dialog(group="Enthalpy sensor in use", enable = fixEnt));
+  parameter Real delTemHis=1 "Delta between the temperature hysteresis high and low limits";
+  parameter Real damLimControllerGain=1 "Gain of damper limit controller";
+  parameter Real modControllerGain=1 "Gain of modulation controller";
```
Use `k` or `kP` or `kMod` or `kPMod` to be clear what gain it is.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,179 @@
+    annotation (Placement(transformation(extent={{-140,70},{-120,90}}),
+        iconTransformation(extent={{-120,50},{-100,70}})));
+  CDL.Interfaces.RealInput VOut(unit="m3/s")
+    "Measured outdoor volumetirc airflow rate [fixme: which quantity attribute should we use? add for all V]"
```
Use `VOut_flow` (we use `_flow` for mass flow rate, volume flow rate, enthalpy flow rate etc.)

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,179 @@
+    "Measured outdoor volumetirc airflow rate [fixme: which quantity attribute should we use? add for all V]"
+    annotation (Placement(transformation(extent={{-140,10},{-120,30}}),
+        iconTransformation(extent={{-120,-10},{-100,10}})));
+  CDL.Interfaces.RealInput VOutMinSet(unit="m3/s")
```
Add `_flow` as above

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,179 @@
+    "Minimum outdoor volumetric airflow rate setpoint"
+    annotation (Placement(transformation(extent={{-140,-10},{-120,10}}),
+        iconTransformation(extent={{-120,-30},{-100,-10}})));
+  CDL.Interfaces.IntegerInput uFreProSta "Freeze protection status"
```
Add `unit="1"`, also for signals below.

milicag replied:
```Integer``` does not seem to have ```unit``` attribute.
```
type Integer
  IntegerType value;
  parameter StringType quantity;
  parameter IntegerType min, max;
  parameter IntegerType start;
  parameter BooleanType fixed;
end Integer;
```

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,179 @@
+    iconTransformation(extent={{100,-30}, {120,-10}})));
+
+  AtomicSequences.EconEnableDisableMultiZone econEnableDisableMultiZone(
+    delEntHis=delEntHis,
```
Most if not all of these assignments should be final.

milicag replied:
Validation tests are looking at some of the outputs of these blocks, which is why I did not protect them. Let me know if there is a suitable workaround.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/Validation/EconomizerMultiZone_Mod_DamLim.mo:
```diff
@@ -0,0 +1,186 @@
+  parameter Real TSup(unit="K", quantity="TermodynamicTemperature")=290
+    "Measured supply air temperature";
+
+  parameter Real minVOutSet(unit="m3/s")=0.71
```
Use `Modelica.SIunits.VolumeFlowRate`

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Types/OperationMode.mo:
```diff
@@ -0,0 +1,9 @@
+    occupied "Occupied Mode",
+    coolDown "Cool-down Mode",
+    setup "Setup Mode",
+    warmup "Warmup Mode",
```
If we use coolDown, we should also use warmUp rather than warmup to be consistent.

milicag commented:
@mwetter I've addressed your comments.

milicag commented:
@JayHuLBL Jianjun, could you maybe review the references to G36 in the sequences I've added in this PR? I believe I've covered all the related sections of the guideline, but an another pair of eyes before the merge is always useful.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/VAVSingleZoneTSupSet.mo:
```diff
@@ -455,61 +453,61 @@ equation
+     </p>
+     <p align=\"center\">
+     <img alt=\"Image of set point reset\"
+     src=\"modelica://Buildings/Resources/Images/Experimental/OpenBuildingControl/ASHRAE/G36/VAVSingleZoneTSupSet.png\"/>
```
@milicag : Please also move the figure

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconDamperPositionLimitsMultiZone_LoopDisable.mo:
```diff
@@ -0,0 +1,183 @@
+    "Maximum increase in airflow volume during the example simulation";
+
+  // Fan Status
+  CDL.Logical.Constant FanStatus(k=false) "Fan is off"
```
`FanStatus` is an instance and hence should be lower case, e.g., use `fanStatus`

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconDamperPositionLimitsMultiZone_LoopDisable.mo:
```diff
@@ -0,0 +1,183 @@
+    "Freeze protection stage 1";
+  parameter Types.FreezeProtectionStage freProStage2 = Types.FreezeProtectionStage.stage2
+    "Freeze protection stage 2";
+  parameter Integer freProStage1Num = Integer(freProStage1)-1
```
This is confusing. Why do you need to subtract 1 here? Can't you subtract it when you assign it in a block, so we have one value for freeze protection stage 1, and not two different values depending on whether the parameter ends in "Num" or not.

milicag replied:
If I don't subtract it, we will have an integer value 1 for stage0, which, I thought might be confusing. Let me know if you'd still like me to use 1 for stage0, 2 for stage1, etc.

mwetter replied:
I see. I committed to this branch a new package
`Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.Constants.FreezeProtectionStages`. Please use this and delete the enumeration. The new package assigns the integer value `0` to `stage0`, etc.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconDamperPositionLimitsMultiZone_VOut_flow.mo:
```diff
@@ -0,0 +1,86 @@
+protected
+  parameter Types.FreezeProtectionStage freProDisabled = Types.FreezeProtectionStage.stage0
+    "Indicates that the freeze protection is disabled";
+  parameter Integer freProDisabledNum = Integer(freProDisabled)-1
```
Please don't subtract one here, see above comment.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconDamperPositionLimitsMultiZone_VOut_flow.mo:
```diff
@@ -0,0 +1,86 @@
+  parameter Types.OperationMode occupied = Types.OperationMode.occupied
+    "Operation mode is \"Occupied\"";
+  parameter Integer occupiedNum = Integer(occupied)
+    "Numerical value for \"Occupied\" operation mode (=1)";
```
Don't use quotes (we also don't use it in other models, and you don't use it on line 41 either.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconEnableDisableMultiZone_FreProSta_ZonSta.mo:
```diff
@@ -0,0 +1,157 @@
+  "Model validates economizer disable for heating zone state and activated freeze protection"
+  extends Modelica.Icons.Example;
+
+  parameter Real TOutCutoff(final unit="K", quantity="TermodynamicTemperature")=297
```
Use Modelica.SIunits.Temperature

milicag replied:
Should I also use Modelica.SIunits.Enthalpy?

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconEnableDisableMultiZone_FreProSta_ZonSta.mo:
```diff
@@ -0,0 +1,157 @@
+    annotation (Placement(transformation(extent={{82,-40},{102,-20}})));
+
+protected
+  CDL.Continuous.Constant TOutBellowCutoff(k=TOutCutoff - 2)
```
`...Below...` has one l

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/EconEnableDisableMultiZone_TOut_hOut.mo:
```diff
@@ -0,0 +1,210 @@
+  "Model validates economizer disable in case outdoor air conditions are above cutoff"
+  extends Modelica.Icons.Example;
+
+  parameter Real TOutCutoff(final unit="K", quantity="TermodynamicTemperature")=297
```
Use `Modelica.SIunits.`

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/Validation/package.mo:
```diff
@@ -1,7 +1,9 @@
 package Validation "Package with validation models"
   extends Modelica.Icons.ExamplesPackage;
 
+
```
Don't add new lines.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/AtomicSequences/package.mo:
```diff
@@ -0,0 +1,7 @@
+within Buildings.Experimental.OpenBuildingControl.ASHRAE.G36;
+package AtomicSequences "Subsequences as defined in guidline G36"
```
Package info is missing

milicag replied:
@milicag: left off here.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,185 @@
+
+  parameter Boolean use_enthalpy = true
+    "Set to true if enthalpy measurement is used in addition to temperature measurement";
+  parameter Real kPIDamLim=1 "Gain of damper limit controller";
```
This should be `kP`, not `kPI`. The gain is for the proportional part, not the integral part. Same for line 7.
Also, lines 8 and 9 should be TiDamLin, not PI

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,185 @@
+    annotation (Placement(transformation(extent={{-140,70},{-120,90}}),
+        iconTransformation(extent={{-120,50},{-100,70}})));
+  CDL.Interfaces.RealInput VOut_flow(unit="m3/s")
+    "Measured outdoor volumetirc airflow rate [fixme: which quantity attribute should we use? add for all V]"
```
Address all fixme for a pull request.
Volumetric has a typo.
For quantity, see Modelica.SIunits.VolumeFlowRate

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,185 @@
+  CDL.Interfaces.IntegerInput uOperationMode "AHU operation mode status signal"
+    annotation (Placement(transformation(extent={{-140,-90},{-120,-70}}),
+      iconTransformation(extent={{-120,-70},{-100,-50}})));
+  CDL.Interfaces.IntegerInput uZoneState "Zone state input"
```
Use 3 letters, e.g., `uZonSta` -- otherwise we will never be consistent in our notation, and searching for results will be harder.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,185 @@
+  outdoor and return air damper positions based on ASHRAE
+  Guidline 36, sections: PART5 N.2.c, N.5, N.6.c, N.7, A.17, N.12.
+  The sequence comprises the following atomic sequences:
+  <code>EconDamperPositionLimitsMultiZone</code>,
```
You need to use the full Modelica name with a hyperlink, such as <a href=\"modelica://Buildings. ....\">Buildings. ...</a>
Otherwise, if we rename models with the rename script, these names will not get updated, and there is no mechanism to test whether all names are valid (for href, we can check that the name is correct when generating the html)

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/Validation/EconomizerMultiZone_Mod_DamLim.mo:
```diff
@@ -0,0 +1,186 @@
+    "Example volumetric airflow setpoint, 15cfm/occupant, 100 occupants";
+  parameter Modelica.SIunits.VolumeFlowRate minVOut_flow=0.705
+    "Minimal measured volumetric airflow";
+  parameter Modelica.SIunits.VolumeFlowRate VOut_flowIncrease=0.03
```
Use the `_flow` word at the end of the variable.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/Validation/EconomizerMultiZone_Mod_DamLim.mo:
```diff
@@ -0,0 +1,186 @@
+  parameter Types.ZoneState deadband = Types.ZoneState.deadband
+    "Zone state is deadband";
+  parameter Integer deadbandNum = Integer(deadband)
+    "Numerical value for deadband zone state (=2)";
```
Don't use "(=2)". The point of enumerations is that one can change it for example by adding another element anywhere in the enumeration. If we end up changing it, we likely won't notice that somewhere in the code there is a comment that says deadband is 2, which will be wrong if a new element is added at the beginning of an enumeration.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/Validation/EconomizerMultiZone_Mod_DamLim.mo:
```diff
@@ -0,0 +1,186 @@
+          fontSize=8,
+          textString="Economizer fully enabled - 
+validate damper position limits")}),
+    Documentation(info="<html>
```
In the future, please don't use spaces in the info section at the start of each line.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/Validation/package.mo:
```diff
@@ -0,0 +1,17 @@
+within Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.CompositeSequences;
+package Validation
```
info section is missing

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/package.mo:
```diff
@@ -0,0 +1,27 @@
+within Buildings.Experimental.OpenBuildingControl.ASHRAE.G36;
+package CompositeSequences "Sequences as defined in G36"
```
info section is missing, and remove a few empty lines

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/IO/Hardware/package.mo:
```diff
@@ -0,0 +1,3 @@
+within Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.IO;
```
info section is missing

milicag replied:
added

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/IO/Hardware/package.mo:
```diff
@@ -0,0 +1,3 @@
+within Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.IO;
+package Hardware
```
info section is missing

milicag replied:
added

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/IO/Software/package.mo:
```diff
@@ -0,0 +1,4 @@
+within Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.IO;
+package Software
```
info section is missing

milicag replied:
added

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/CompositeSequences/Validation/EconomizerMultiZone_Disable.mo:
```diff
@@ -0,0 +1,175 @@
+  "Validation model for disabling the multizone VAV AHU economizer modulation and damper position limit control loops"
+  extends Modelica.Icons.Example;
+
+  parameter Real TOutCutoff(unit="K", quantity="TermodynamicTemperature")=297
```
Use Modelica.SIunits.

milicag commented:
@mwetter I added ```Zone State``` to ```Constants``` and incorporated the new constants packages in all blocks and models. For some reason there is a merge conflict since ```VAVSingleZoneTSupSet.mo``` still lives in ```G36``` package.

milicag commented:
@mwetter I resolved the post-merge issues and updated blocks/models based on a change in CDL packages. Please let me know if you'd like any further edits.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconDamperPositionLimitsMultiZone.mo:
```diff
@@ -0,0 +1,312 @@
+        iconTransformation(extent={{100,-50},{120,-30}})));
+
+  CDL.Continuous.LimPID damLimController(
+    Ti=TiDamLim,
```
Please make these final as otherwise users can overwrite them which is not the intention. And conversely, if we release the code and afterwards make them final, then the code will not be backwards compatible.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconEnableDisableMultiZone.mo:
```diff
@@ -0,0 +1,348 @@
+    "OA enthalpy high limit cutoff. For differential enthalpy use return air enthalpy measurement"
+    annotation (Placement(transformation(extent={{-220,130},{-180,170}}),iconTransformation(extent={{-120,30},{-100,50}})));
+  CDL.Interfaces.RealInput uOutDamPosMin(min=0, max=1)
+    "Minimum outdoor air damper position, from EconDamperPositionLimitsMultiZone sequence"
```
Please only use model names if you can make a hyperlink. Here, simply write in words where to get signal from. Otherwise, if we rename `EconDamperPositionLimitsMultiZone` with a conversion script, then this text will not be updated and hence references a block that no longer exists. As we have 100s of files, detecting such broken documentation is very hard, hence we want to avoid the risk of breaking it.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconEnableDisableMultiZone.mo:
```diff
@@ -0,0 +1,348 @@
+  CDL.Logical.Hysteresis hysOutEnt(final uLow=uEntHigLimCutLow, final uHigh=uEntHigLimCutHig) if use_enthalpy
+    "Outdoor air enthalpy hysteresis for both fixed and differential enthalpy cutoff conditions"
+    annotation (Placement(transformation(extent={{-100,160},{-80,180}})));
+  CDL.Logical.Switch OutDamSwitch "Set maximum OA damper position to minimum at disable (after time delay)"
```
use lower case for instance name

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconEnableDisableMultiZone.mo:
```diff
@@ -0,0 +1,348 @@
+  connect(add2.y, hysOutEnt.u) annotation (Line(points={{-119,170},{-102,170}}, color={0,0,127}));
+  connect(hysOutTem.y, nor1.u1) annotation (Line(points={{-79,250},{-60,250},{-60,210},{-42,210}},
+    color={255,0,255}));
+  if use_enthalpy then
```
this is not allowed in CDL. Only connect statements are allowed.
However, if `hysOutEnt` is removed based on the parameter value, then the connect statement is also removed. Hence, make sure that either `hysOutEnt` or `entSubst` is removed.

milicag replied:
Thanks, good hint. I thought it wouldn't work without it, but it does.

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconModulationMultiZone.mo:
```diff
@@ -0,0 +1,197 @@
+    "Damper position is linearly proportional to the control signal between signal limits"
+    annotation (Placement(transformation(extent={{60,60},{80,80}})));
+  CDL.Continuous.LimPID damPosController(
+    controllerType=Buildings.Experimental.OpenBuildingControl.CDL.Types.SimpleController.PI,
```
make assignments final

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,185 @@
+  Atomic.EconEnableDisableMultiZone ecoEnaDis(
+    final delEntHis=delEntHis,
+    final delTemHis=delTemHis,
+    use_enthalpy=use_enthalpy) "Multizone VAV AHU economizer enable/disable sequence"
```
make final

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,185 @@
+    use_enthalpy=use_enthalpy) "Multizone VAV AHU economizer enable/disable sequence"
+    annotation (Placement(transformation(extent={{0,-40},{20,-20}})));
+  Atomic.EconDamperPositionLimitsMultiZone ecoDamLim(
+    retDamPhyPosMax=0.9,
```
Make final, or if the user should adjust them, propagate them to top-level parameters

mwetter commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/EconomizerMultiZone.mo:
```diff
@@ -0,0 +1,185 @@
+  parameter Modelica.SIunits.Temperature delTemHis=1
+    "Delta between the temperature hysteresis high and low limits";
+
+public
```
Public instances must come before protected

milicag commented:
@mwetter I've completed this iteration. I've propagated physical damper position limits, since the user likely needs to access those values at some point (e.g. commissioning). I also really like G36.Constants as opposed to G36.Types.

mwetter commented:
This is ready to merge when all tests pass.

mwetter commented:
@milicag : The regression tests fail because `EconomizerMultiZone_Mod_DamLim` and `EconomizerMultiZone_Disable` produce results that are different from the stored reference results. I don't think my changes above that replaced the `Logical.Equal` block with two inequalities, and that changed comparison of integers after conversion to reals, caused this change.
Essentially, rather than testing something like `Real(m) <= 1` where `m` is an integer, I changed to tests to `Real(m) <= 1.5` to be sure not to encounter any rounding errors.

Can you please verify and commit the needed changes in the model or the reference results (whichever is wrong). You can run the broken tests with
```
cd Buildings
../bin/runUnitTests.py -s Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.Composite
```

milicag commented:
@mwetter You are right, the change in reference results has nothing to do with your changes. As I've propagated the physical damper limit parameters, I've changed the values to 0 and 1, whereas the previous test was run with 0.1 and 0.9 for return air damper. I've updated the results using current default values.
